core/trino-main/src/main/java/io/trino/sql/planner/AdaptivePlanner.java

```diff
@@ -0,0 +1,438 @@
+                subPlans,
+                // We do not change the fragment ids which have no changes and are not downstream of the
+                // changed plan nodes. This optimization is done to avoid unnecessary stage restart due to speculative
+                // execution.
+                getUnchangedFragments(adaptivePlan, optimizationResult.changedPlanNodes(), fragmentIdentifierToId),
+                false,
+                warningCollector);
+    }
+
+    private AdaptivePlanOptimizer.Result optimizePlan(
+            PlanNode plan,
+            SymbolAllocator symbolAllocator,
+            TableStatsProvider tableStatsProvider,
+            StageRuntimeInfoProvider stageRuntimeInfoProvider,
+            PlanNodeIdAllocator idAllocator)
+    {
+        AdaptivePlanOptimizer.Result result = new AdaptivePlanOptimizer.Result(plan, Set.of());
+        ImmutableSet.Builder<PlanNodeId> changedPlanNodes = ImmutableSet.builder();
+        for (AdaptivePlanOptimizer optimizer : planOptimizers) {
+            result = optimizer.optimizeAndReturnChanges(
+                    result.plan(),
+                    new PlanOptimizer.Context(
+                            session,
+                            symbolAllocator.getTypes(),
+                            symbolAllocator,
+                            idAllocator,
+                            warningCollector,
+                            planOptimizersStatsCollector,
+                            tableStatsProvider,
+                            stageRuntimeInfoProvider));
+            changedPlanNodes.addAll(result.changedPlanNodes());
+        }
+        return new AdaptivePlanOptimizer.Result(result.plan(), changedPlanNodes.build());
+    }
+
+    private PlanNode addAdaptivePlanNode(
+            PlanNodeIdAllocator idAllocator,
+            PlanNode initialPlan,
+            PlanNode optimizedPlanNode,
+            TypeProvider types,
+            Set<PlanNodeId> changedPlanNodes)
+    {
+        if (changedPlanNodes.contains(optimizedPlanNode.getId())) {
+            return new AdaptivePlanNode(
+                    idAllocator.getNextId(),
+                    initialPlan,
+                    getFilteredSymbols(initialPlan, types),
+                    optimizedPlanNode);
+        }
+
+        verify(initialPlan.getSources().size() == optimizedPlanNode.getSources().size());
+        ImmutableList.Builder<PlanNode> sources = ImmutableList.builder();
+        for (int i = 0; i < initialPlan.getSources().size(); i++) {
+            PlanNode initialSource = initialPlan.getSources().get(i);
+            PlanNode optimizedSource = optimizedPlanNode.getSources().get(i);
+            sources.add(addAdaptivePlanNode(idAllocator, initialSource, optimizedSource, types, changedPlanNodes));
+        }
+        return optimizedPlanNode.replaceChildren(sources.build());
+    }
+
+    private Map<FragmentIdentifier, PlanFragmentId> getUnchangedFragments(
```

Comment:
This is essentially looking for a complete paths from root to changed nodes. Could we maybe use some Guava for traversing and accumulating?